### ISSUE TITLE
Add Automatic-Module-Name to MANIFEST.MF

### ISSUE THREAD

**Issue by dbwiddis** [oshi/oshi#1243]
This is apparently a minimal good idea to support modules without changing anything else.

I believe it's as simple as configuring the `maven-jar-plugin` entry in the POM:

```
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-jar-plugin</artifactId>
        <configuration>
          <archive>
            <manifestEntries>
              <Automatic-Module-Name>${project.groupId}.${project.artifactId}</Automatic-Module-Name>
            </manifestEntries>
          </archive>
        </configuration>
      </plugin>
```

But I'd really prefer someone more familiar with Java 9+ confirm this, and maybe even submit the PR.

**Comment by dbwiddis** [comment]
OK, so the above won't work.
 - Module names aren't supposed to have hyphens, so `oshi-core` is off-limits
 - The module is supposed to match the highest level package name, which in our case is simply `oshi`.   But claiming ownership of all `oshi.*` packages in Java 9+ forever is probably not a good thing.  `com.github.oshi` seems appropriate, but does that mean I need to retitle all the packages in the project?

**Comment by dbwiddis** [comment]
Leaning toward the above with this modification:
Parent:
```
<Automatic-Module-Name>${javaModuleName}</Automatic-Module-Name>
```
Each artifact defines the module name in its properties.  Probably `com.github.oshi` for `oshi-core` and `com.github.oshi.demo` for `oshi-demo`.  But we don't need module definitions for the `-dist` or even `-parent` artifacts.  And I don't know how shaded jar will work with the modules...

**Comment by dbwiddis** [comment]
OK, so I've come to some conclusions here.
1. The shaded jar shouldn't have a module name. Uber-jars don't play nice with modules.
2. Rather than defining the configuration in the parent, the `oshi-core` and `oshi-demo` modules can place that definition in their own poms, eventually.
3. The module names should be `com.github.oshi` and `com.github.oshi.demo` respectively.  However, that also means the top-level package names should also prepend `com.github.` to their existing names.  This would require everyone upgrading to change every import in their code.  I'm not too fond of doing this just to enable this feature when we're not using any modules in OSHI to begin with.  Maybe if the project gets its own domain name someday I might do something non-github-ish.  

Open to arguments otherwise, but for now Java 9+ people will have to deal with whatever the automatic resolution provides.

**Comment by dbwiddis** [comment]
After sleep and coffee, I'm going to go ahead and add the "com.github" module names just for the core and demo artifacts using some pom "append" tricks I just learned, but leave the project's package structure the same. It's not any worse than the current default "oshi.core" module mismatching its contents, and better long-term.

Should make a few Java 9 folks happy, at least. :)

**Comment by hazendaz** [comment]
@dbwiddis We have this setup on mybatis.  See here https://github.com/mybatis/parent/blob/master/pom.xml.  You probably could copy that ,pretty close to what you are coming up with.  In that case, we actually do this in all modules without much concern of it being a pom or distro or whatever.    Since the parent defines it, absense in lower used modules would end up defined as parent unless otherwise added.  When we did this, I think we did it all pretty quickly to all modules at once.

**Comment by hazendaz** [comment]
Guess I was too slow ;) Looks good.  I don't recall your structure but if you define it once in parent then just define properly in other poms, you can bypass setting up jar plugin multiple times.

**Comment by dbwiddis** [comment]
Yep, but then I'd end up with a module name for shaded jar (not a good idea, I've read) and for the "dist" (which is just a bunch of zip files).  Probably harmless other than the mental effort in choosing a module name.

I think what I did required the minimum amount of fuss... just append one line in the two projects that I actually want that module name.  And it's done.

Renaming packages is a bigger question, and not one I plan to tackle any time soon. :)

**Comment by dbwiddis** [comment]
And it may not be so simple.  The [Appveyor JDK11 build](https://ci.appveyor.com/project/dbwiddis/oshi/build/job/1agm8lghtfk8eiel) can't find packages in the core artifact when building the shaded jar.  JDK8 doesn't complain but perhaps JDK11 doesn't like this band-aid fix...

**Comment by hazendaz** [comment]
I think there is a bug in shade plugin latest.  Maven is working on fix due to a regression.  Not sure if related but search on maven nabble and look at developer channel to find info on it.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by dbwiddis** [comment]
OK, thanks for that pointer.  I'll let this PR sit for a bit and maybe try the simpler properties-based implementation just to reduce pom-spam.

**Comment by dbwiddis** [comment]
So, I played with a few things today.  Properties-based implementation doesn't help.  I thought "hey, let me just actually create the `module-info` files and do it right" and learned about multi-Jar releases and everything.  I think I can make that change.... soon.   Problem is, we have one dependency (code-assert) that does not yet have a defined module name and the auto-generated fallback is.... `code.assert`.  Which is an illegal module name so I can't import it.     Put in a PR at that project to get a module name and maybe I can get set up for a multi-jar/modularized build.

**Comment by dbwiddis** [comment]
Well, that's been an interesting week. I feel like I'm an expert on JPMS now after all my various attempts to actually set up modules on OSHI. Or a multi-release Jar.  Trying to explicitly do it failed because of the code-assert dependency name resolution, and various attempts to generate a name, or even insert a module-info class using ModiTect, all failed.  (Actually ModiTect flipped the problem... worked brilliantly on JDK 11 but failed on JDK8 due to Java9 classes.).  I really like ModiTect and will use it in the future when we modularize.

Ah, but it seems the fix to the javadoc plugin's failure to find documents was to add this to the config in the shaded-jar pom:
`<sourcepath>src/main/java</sourcepath>`

I don't know why it works (or why it doesn't work without it) but it looks like the problem is resolved.  I think there's an outside chance this may also fix #1145, but I need to cut a release to test that.
